r/R/dplyr-summarize.R

```diff
@@ -28,14 +28,107 @@ summarise.arrow_dplyr_query <- function(.data, ..., .engine = c("arrow", "duckdb
     dplyr::group_vars(.data) # vars needed for grouping
   ))
   .data <- dplyr::select(.data, vars_to_keep)
-
   if (match.arg(.engine) == "duckdb") {
     dplyr::summarise(to_duckdb(.data), ...)
-  } else {
-    if (query_on_dataset(.data)) {
-      not_implemented_for_dataset("summarize()")
+  } else if (isTRUE(getOption("arrow.summarize", FALSE))) {
+    # Try stuff, if successful return()
+    out <- try(do_arrow_summarize(.data, ...), silent = TRUE)
+    if (inherits(out, "try-error")) {
+      return(abandon_ship(call, .data, format(out)))
+    } else {
+      return(out)
     }
+  } else {
+    # If unsuccessful or if option not set, do the work in R
     dplyr::summarise(dplyr::collect(.data), ...)
   }
 }
 summarise.Dataset <- summarise.ArrowTabular <- summarise.arrow_dplyr_query
+
+do_arrow_summarize <- function(.data, ...) {
```

Comment:
Handle the `.groups` argument to `dplyr::summarise()`:
```suggestion
do_arrow_summarize <- function(.data, ..., .groups = NULL) {
  if (!is.null(.groups)) {
    abort("`summarize()` with `.groups` argument not supported in Arrow")
  }
```